src/target/arc32.c

```diff
@@ -472,28 +518,53 @@ int arc32_dcache_flush(struct target *target)
 
 	/* Store current value of DC_CTRL */
 	CHECK_RETVAL(arc_jtag_read_aux_reg_one(&arc32->jtag_info, AUX_DC_CTRL_REG, &dc_ctrl_value));
-
 	/* Set DC_CTRL invalidate mode to flush (if not already set) */
 	has_to_set_dc_ctrl_im = (dc_ctrl_value & DC_CTRL_IM) == 0;
 	if (has_to_set_dc_ctrl_im) {
 		value = dc_ctrl_value | DC_CTRL_IM;
 		CHECK_RETVAL(arc_jtag_write_aux_reg_one(&arc32->jtag_info, AUX_DC_CTRL_REG, value));
 	}
-
 	/* Flush D$ */
 	value = DC_IVDC_INVALIDATE;
 	CHECK_RETVAL(arc_jtag_write_aux_reg_one(&arc32->jtag_info, AUX_DC_IVDC_REG, value));
-
 	/* Restore DC_CTRL invalidate mode (even of flush failed) */
 	if (has_to_set_dc_ctrl_im) {
 	    CHECK_RETVAL(arc_jtag_write_aux_reg_one(&arc32->jtag_info, AUX_DC_CTRL_REG, dc_ctrl_value));
 	}
-
 	arc32->dcache_flushed = true;
 
 	return ERROR_OK;
 }
 
+static int arc32_l2cache_flush(struct arc32_common *arc32)
+{
+	uint32_t value = L2_FLL2_FLUSH;
+
+	if (!arc32->has_l2cache || arc32->l2cache_flushed)
+	    return ERROR_OK;
+
+	LOG_DEBUG("Flushing L2$.");
+
+	/* Flush L2 cache */
+	CHECK_RETVAL(arc_jtag_write_aux_reg_one(&arc32->jtag_info, AUX_L2_FLL2_REG, value));
+	/* Wait until flush operation ends */
+	do {
+	    CHECK_RETVAL(arc_jtag_read_aux_reg_one(&arc32->jtag_info, AUX_L2_CTRL_REG, &value));
```

Comment:
I would add a LOG_DEBUG message here, in case there is some bug and this loop turns into infinite loop.